src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/DeterministicShuffleShardSyncLeaderDecider.java

```diff
@@ -0,0 +1,184 @@
+    private static final Log LOG = LogFactory.getLog(DeterministicShuffleShardSyncLeaderDecider.class);
+    private static final int LEADER_DECIDER_THREAD_COUNT = 1;
+    private static final long ELECTION_INITIAL_DELAY_MILLIS = 60 * 1000;
+    private static final long ELECTION_SCHEDULING_INTERVAL = 5 * 60 * 1000;
+    private static final int AWAIT_TERMINATION_MILLIS = 5000;
+
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+    private final KinesisClientLibConfiguration config;
+    private final ILeaseManager<KinesisClientLease> leaseManager;
+    private final int numPeriodicShardSyncWorkers;
+    private final ScheduledExecutorService leaderElectionThreadPool;
+
+    private volatile Set<String> leaders;
+
+    /**
+     * Create an instance using the default periodic shard sync worker count (1).
+     * This constructor is package-private to ensure external users use the public
+     * constructor and are thereby aware of the number of periodic shard sync workers
+     * they want in their application.
+     *
+     * @param config       KinesisClientLibConfiguration instance
+     * @param leaseManager LeaseManager instance.
+     */
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager) {
+        this(config, leaseManager, PERIODIC_SHARD_SYNC_MAX_WORKERS_DEFAULT);
+    }
+
+    /**
+     * Create an instance overriding the default periodic shard sync worker count.
+     * This is public for use outside KCL.
+     *
+     * @param config                      KinesisClientLibConfiguration instance
+     * @param leaseManager                LeaseManager instance.
+     * @param numPeriodicShardSyncWorkers Max number of periodic shard sync workers.
+     */
+    public DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, int numPeriodicShardSyncWorkers) {
+        this(config, leaseManager, Executors.newScheduledThreadPool(LEADER_DECIDER_THREAD_COUNT),
+                numPeriodicShardSyncWorkers);
+    }
+
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, ScheduledExecutorService leaderElectionThreadPool,
+            int numPeriodicShardSyncWorkers) {
+        this.config = config;
+        this.leaseManager = leaseManager;
+        this.leaderElectionThreadPool = leaderElectionThreadPool;
+        this.numPeriodicShardSyncWorkers = numPeriodicShardSyncWorkers;
+    }
+
+    /*
+     * Shuffles the leases deterministically and elects numPeriodicShardSyncWorkers number of workers
+     * as leaders (workers that will perform shard sync).
+     */
+    private void electLeaders() {
+        try {
+            LOG.debug("Started leader election: " + System.currentTimeMillis());
```

Comment:
We generally use 'Instant' to capture time in KCL

Reply:
Changed.